src/nvim/ex_getln.c

```diff
@@ -2306,6 +2313,267 @@ static int empty_pattern(char_u *p)
+  int retv = rename_buffer("[Preview]");
+  aucmd_restbuf(&aco);
+
+  if (retv == FAIL) {
+    return NULL;
+  }
+
+  // Temporarily switch to preview buffer to set it up for previewing.
+  aucmd_prepbuf(&aco, cmdpreview_buf);
+  buf_clear();
+  curbuf->b_p_ma = true;
+  curbuf->b_p_ul = -1;
+  curbuf->b_p_tw = 0;  // Reset 'textwidth' (was set by ftplugin)
+  aucmd_restbuf(&aco);
+  cmdpreview_bufnr = cmdpreview_buf->handle;
+
+  return cmdpreview_buf;
+}
+
+/// Open command preview window if it's not already open.
+/// Returns to original window after opening command preview window.
+///
+/// @param cmdpreview_buf Pointer to command preview buffer
+///
+/// @return Pointer to command preview window if succeeded, NULL if failed.
+static win_T *cmdpreview_open_win(buf_T *cmdpreview_buf)
+{
+  win_T *save_curwin = curwin;
+  bool win_found = false;
+
+  // Try to find an existing preview window.
+  FOR_ALL_WINDOWS_IN_TAB(wp, curtab) {
+    if (wp->w_buffer == cmdpreview_buf) {
+      win_enter(wp, false);
+      win_found = true;
+      break;
+    }
+  }
+
+  // If an existing window is not found, create one.
+  if (!win_found && win_split((int)p_cwh, WSP_BOT) == FAIL) {
+    return NULL;
+  }
+
+  win_T *preview_win = curwin;
+  Error err = ERROR_INIT;
+
+  // Switch to preview buffer
+  try_start();
+  int result = do_buffer(DOBUF_GOTO, DOBUF_FIRST, FORWARD, cmdpreview_buf->handle, 0);
+  if (try_end(&err) || result == FAIL) {
+    api_clear_error(&err);
+    return NULL;
+  }
+
+  curwin->w_p_cul = false;
+  curwin->w_p_cuc = false;
+  curwin->w_p_spell = false;
+  curwin->w_p_fen = false;
+
+  win_enter(save_curwin, false);
+  return preview_win;
+}
+
+/// Closes any open command preview windows.
+static void cmdpreview_close_win(void)
+{
+  buf_T *buf = cmdpreview_bufnr ? buflist_findnr(cmdpreview_bufnr) : NULL;
+  if (buf != NULL) {
+    close_windows(buf, false);
+  }
+}
+
+/// Show 'inccommand' preview. It works like this:
+///    1. Store current undo information so we can revert to current state later.
+///    2. Execute the preview callback with the parsed command, preview buffer number and preview
+///       namespace number as arguments. The preview callback sets the highlight and does the
+///       changes required for the preview if needed.
+///    3. Preview callback returns 0, 1 or 2. 0 means no preview is shown. 1 means preview is shown
+///       but preview window doesn't need to be opened. 2 means preview is shown and preview window
+///       needs to be opened if inccommand=split.
+///    4. Use the return value of the preview callback to determine whether to
+///       open the preview window or not and open preview window if needed.
+///    5. If the return value of the preview callback is not 0, update the screen while the effects
+///       of the preview are still in place.
+///    6. Revert all changes made by the preview callback.
+static void cmdpreview_show(CommandLineState *s)
```

Comment:
A key feature of 'inccommand' `:substitute` is that it skips the preview if the preview takes too long: https://github.com/neovim/neovim/blob/46536f53e82967dcac8d030ee3394cdb156f9603/src/nvim/ex_cmds.c#L4430-L4432

That should also be built into the handling of user-defined cmd preview.

Reply:
Another important optimization was to stop looking for matches if additional ones would not be shown to the user (either the preview buffer is full and icm=split, or they're out of the viewport of the current buffer with icm=nosplit). This PR is too complicated for me, so I could not discern if that was implemented, but I figured I'd mention it (ofc it's most helpfull in large files).

Reply:
> A key feature of 'inccommand' `:substitute` is that it skips the preview if the preview takes too long: https://github.com/neovim/neovim/blob/46536f53e82967dcac8d030ee3394cdb156f9603/src/nvim/ex_cmds.c#L4430-L4432
> 
> That should also be built into the handling of user-defined cmd preview.

How should we implement that though? Or should we leave that to the preview callback?